### PR TITLE
FIX: Allow KNN with missing values

### PR DESCRIPTION
**Pull request by david-cortes-intel**
## Description

KNN algorithms on sklearn support data with missing values, but sklearnex makes a check for 'all finite'. This PR makes it fall back to sklearn when receiving data that has missing values.

Disclaimer: I'm not 100% sure about whether oneDAL supports KNN with missing values or not.

---

<details>
<summary>Checklist:</summary>

**Completeness and readability**

- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.

</details>

### CODE REVIEW

File: sklearnex/neighbors/common.py [hunk 1]
```diff
@@ -254,6 +257,21 @@ def _onedal_supported(self, device, method_name, *data):
                 ]
             ):
                 return patching_status
+        X = data[0]
+        if X is not None and not sp.issparse(X):
+            xp, _ = get_namespace()
+            try:
+                has_nan = xp.any(xp.isnan(X))
```
Comment (icfaust) [inline sklearnex/neighbors/common.py]: This should instead be done with the optimized `assert_all_finite` as this has performance implications and impacts follow-on finite checks. I recommend looking at the ensemble algos and @Alexsandruss 's solution for this.

Comment (david-cortes-intel) [reply sklearnex/neighbors/common.py]: Switch to use the optimized internal version.

**Comment by icfaust** [comment]
scikit-learn/scikit-learn#25319 @david-cortes-intel . Can you detail the missing value use case with an example? As of now we support these metrics: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/sklearnex/neighbors/common.py#L221 which does not include `nan_euclidean` which would be the use case of this missing value support.

**Comment by david-cortes-intel** [comment]
> [scikit-learn/scikit-learn#25319](https://github.com/scikit-learn/scikit-learn/issues/25319) @david-cortes-intel . Can you detail the missing value use case with an example? As of now we support these metrics: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/sklearnex/neighbors/common.py#L221 which does not include `nan_euclidean` which would be the use case of this missing value support.

It still fails this test:
https://github.com/scikit-learn/scikit-learn/blob/4a10d0ed8d85e6ed24a647bd28a65c0c64b101ef/sklearn/neighbors/tests/test_neighbors.py#L2356

**Comment by icfaust** [comment]
@david-cortes-intel then I think there is a bug in the metrics check. I think that should be the avenue taken to fix the issue.

**Comment by icfaust** [comment]
Just wanted to note that I know that its a bit convoluted and annoying on the neighbors side of things here, I just want to make sure that if we can solve it without a separate finite check, then we should do that.

**Comment by david-cortes-intel** [comment]
> @david-cortes-intel then I think there is a bug in the metrics check. I think that should be the avenue taken to fix the issue.

Switched it to use the metric checks instead.
